File: proxy/response.go
package proxy

import (
  pb "github.com/pantonshire/goldcrest/protocol"
  "github.com/pantonshire/goldcrest/proxy/model"
  "google.golang.org/grpc/metadata"
)

func serError(err error) (*pb.Error, metadata.MD) {
  if proxyError, ok := err.(proxyError); ok {
    return proxyError.ser()
  }
  return nil, nil
}

func generateTweetResponse(generator func() (model.Tweet, metadata.MD, error)) (*pb.TweetResponse, metadata.MD, error) {
  tweet, meta, err := generator()
  if err != nil {
    if errMsg, errMeta := serError(err); errMsg != nil {
      return &pb.TweetResponse{Response: &pb.TweetResponse_Error{Error: errMsg}}, metadata.Join(meta, errMeta), nil
    }
    return nil, nil, err
  }
  return &pb.TweetResponse{Response: &pb.TweetResponse_Tweet{Tweet: serTweet(tweet)}}, meta, nil
}

func generateTweetsResponse(generator func() (model.Timeline, metadata.MD, error)) (*pb.TweetsResponse, metadata.MD, error) {
  tweets, meta, err := generator()
  if err != nil {
    if errMsg, errMeta := serError(err); errMsg != nil {
      return &pb.TweetsResponse{Response: &pb.TweetsResponse_Error{Error: errMsg}}, metadata.Join(meta, errMeta), nil
    }
    return nil, nil, err
  }
  return &pb.TweetsResponse{Response: &pb.TweetsResponse_Tweets{Tweets: serTimeline(tweets)}}, meta, nil
}

func generateUserResponse(generator func() (model.User, metadata.MD, error)) (*pb.UserResponse, metadata.MD, error) {
  user, meta, err := generator()
  if err != nil {
    if errMsg, errMeta := serError(err); errMsg != nil {
      return &pb.UserResponse{Response: &pb.UserResponse_Error{Error: errMsg}}, metadata.Join(meta, errMeta), nil
    }
    return nil, nil, err
  }
  return &pb.UserResponse{Response: &pb.UserResponse_User{User: serUser(user)}}, meta, nil
}

File: client/rust/src/response.rs
use crate::twitter1;

pub trait Response<T> {
    fn response_result(self) -> Option<Result<T, twitter1::Error>>;
}

impl Response<twitter1::Tweet> for twitter1::TweetResponse {
    fn response_result(self) -> Option<Result<twitter1::Tweet, twitter1::Error>> {
        Some(match self.response? {
            twitter1::tweet_response::Response::Tweet(msg) => Ok(msg),
            twitter1::tweet_response::Response::Error(err) => Err(err),
        })
    }
}

impl Response<twitter1::Tweets> for twitter1::TweetsResponse {
    fn response_result(self) -> Option<Result<twitter1::Tweets, twitter1::Error>> {
        Some(match self.response? {
            twitter1::tweets_response::Response::Tweets(msg) => Ok(msg),
            twitter1::tweets_response::Response::Error(err) => Err(err),
        })
    }
}

impl Response<twitter1::User> for twitter1::UserResponse {
    fn response_result(self) -> Option<Result<twitter1::User, twitter1::Error>> {
        Some(match self.response? {
            twitter1::user_response::Response::User(msg) => Ok(msg),
            twitter1::user_response::Response::Error(err) => Err(err),
        })
    }
}
